### PR TITLE
Add support for hooking logs from app-services into the log system

### PR DESCRIPTION
**Pull request by thomcc**
This adds a new component that allows logs emitted by rust code from https://github.com/mozilla/application-services to be directed to the `Log` object in the base component. Previously, they were always written to logcat.

It's worth noting that this only works if you're using a [megazord](https://mozilla.github.io/application-services/docs/applications/consuming-megazord-libraries.html) build (Unfortunately, it's more or less impossible for us to make it work outside a megazord). Code running outside a megazord will continue to use logcat (realistically, we don't expect to ship non-megazords, so this basically means the android-components samples and the like).

I avoided doing anything fancy in `Log` to let it know about this type, although I could see an argument for having it be aware to automatically set the max level of RustLog or something, IDK (I don't think we want a hard dependency though, so there would need to be some annoying indirection that I punted on).

Anyway, I didn't really know where to put this. It's a lot less general purpose than the other stuff in `support`, but it doesn't really fit anywhere else either. LMK, I'm open to move it.

### Note:

This won't work until this is included in an app services release, which hasn't happened yet because it hasn't landed in app services yet (https://github.com/mozilla/application-services/pull/472).

Once it does I'll update the versions and add release notes.

### CODE REVIEW

**Comment by thomcc** [comment]
@csadilek This is no longer blocked and should be landable. There's an issue that means we'll have to point things at nick's bintray for a-s things, the forseeable future.

File: build.gradle [hunk 1]
```diff
@@ -40,6 +40,14 @@ allprojects {
...
             url "https://download.servo.org/nightly/maven"
         }
+
+        // Application services is currently published out of nalexander's bintray for reasons.
+        // FIXME: Remove after https://github.com/mozilla/application-services/issues/252
+        // is resolved.
+        maven {
+            name 'appservices'
+            url 'https://dl.bintray.com/ncalexander/application-services'
+        }
```
Comment (pocmo) [inline build.gradle]: Will we synchronize the artifact from there to jcenter like the others?

Comment (thomcc) [reply build.gradle]: I don't fully understand the question. If I'm being honest, I was instructed to do this by @ncalexan, who hopefully does.

Comment (ncalexan) [reply build.gradle]: It looks like we can just roll back https://github.com/mozilla-mobile/android-components/pull/1765/commits/eabc299a0f4e2bfa40093fda6c580413d9ba9dd1 'cuz `org.mozilla.appservices:rustlog` made it to jcenter: https://jcenter.bintray.com/org/mozilla/appservices/rustlog/.  I don't understand why it told me it couldn't be mirrored... maybe I did it without realizing it the first time?

Comment (thomcc) [reply build.gradle]: That commit is gone now, and things work. Resolving.

File: .buildconfig.yml [hunk 2]
```diff
@@ -192,6 +192,10 @@ projects:
     path: components/support/utils
     description: 'A collection of generic helper classes.'
     publish: true
+  support-rustlog:
+    path: components/support/rustlog
+    description: 'A bridge allowing log messages from rust code to be sent to the the log system in support-base'
```
Comment (csadilek) [inline .buildconfig.yml]: nit: 'the the'

File: components/support/rustlog/src/main/java/mozilla/components/support/rustlog/RustLog.kt [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+package mozilla.components.support.rustlog
+
+import mozilla.appservices.rustlog.LogLevelFilter
+import mozilla.appservices.rustlog.RustLogAdapter
+import mozilla.components.support.base.log.Log
+
+object RustLog {
+
+    /**
+     * Enable the rust log adapter. This does (almost) nothing if you are not
```
Comment (csadilek) [inline components/support/rustlog/src/main/java/mozilla/components/support/rustlog/RustLog.kt]: This note (almost nothing) about megazord wasn't clear to me right away (before Nick explained it to me). Can we clarify this or simply write that it's only supported in Megazord for now?

File: docs/changelog.md [hunk 4]
```diff
@@ -38,6 +39,28 @@ permalink: /changelog/
...
+  // Note: By default this is enabled at level DEBUG, which can be adjusted.
+  // (It is recommended you do this for performance if you adjust
+  // `Log.logLevel`).
+  RustLog.setMaxLevel(Log.Priority.INFO)
+  // You can also enable "trace logs", which may include PII
+  // (but can assist debugging) as follows. It is recommended
+  // you not do this in builds you distribute to users.
+  RustLog.setMaxLevel(Log.Priority.DEBUG, true)
+  ```
+  * This is pointless to do when not using a megazord (in which case logs from these libraries will be written directly to logcat automatically).
```
Comment (csadilek) [inline docs/changelog.md]: Same as above. A clarification would be nice. Why?

**Comment by thomcc** [comment]
I added clarification. It's unfortunate, since ideally megazording would just be an optimization, but it actually has semantic ramifications due to each `.so`'s internal variables (such as the global current logger inside the log crate) being hidden from other .so files.

**Comment by csadilek** [comment]
@thomcc thanks! 🚢

**Pull request merged**